### PR TITLE
Impression Tracking Server

### PR DESCRIPTION
**Pull request by scbedd**
ASP.NET Core site. Tested with AppServices + Local.

* Services pixel requests with in-mem cache deduplication of addresses

@mikeharder I'm updating the readmes and making a pass over the generated files to ensure the trailing newline is present.
@Azure/azure-sdk-eng

### CODE REVIEW

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            if (img == null)
+            {
+                img = System.IO.File.ReadAllBytes(imgPath);
+            }
+
+            return img;
+        }
+
+        private string getRequestAddress()
```
Comment (mikeharder) [inline web/pixel-server/PixelServer/Controllers/TrackingController.cs]: I don't think there's any reason to convert to string.  Just use the `IPAddress` object as the cache key.

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        public async Task<ActionResult> Get(string path)
+        {
+            await trackEventAsync(path);
+
+            return File(getCachedImage(), "image/png");
+        }
+
+        private async Task trackEventAsync(string path)
+        {
+            if (telemetry == null)
```
[the remark this answers is not in the document]
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Good though. Done. Testing now.

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs (hunk 1 above)
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Ohhh. Fair! I'll check it.

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.Extensions.Primitives;
+
+namespace PixelServer.Controllers
+{
+    [Route("api/impressions")]
+    [ApiController]
+    public class TrackingController : ControllerBase
+    {
+        // used for manual logs to application insights
+        private static TelemetryClient telemetry;
```
[the remark this answers is not in the document]
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: If I swap to DI injected, can't be readonly. Swapping over to just `private`

[the remark this answers is not in the document]
Comment (mikeharder) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Readonly fields can be assigned in constructor, just not assigned later.

[the remark this answers is not in the document]
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Yeah. Correct. I had it on static as well which was the source of the error. Done.

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        /// Currently the only entrypoint, services a 1 pixel image while recording that there was an impression for the specific path.
+        /// 
+        /// Uses in-memory cache of afore-seen IPs to recognize a "duplicate" impression. The addresses themselves are not recorded.
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        [HttpGet]
+        public async Task<ActionResult> Get(string path)
+        {
+            await trackEventAsync(path);
```
[the remark this answers is not in the document]
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Tried this. It's currently set to `fire and forget` method.

[the remark this answers is not in the document]
Comment (scbedd) [reply web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Swapped it to inline. TEsting.

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+using Microsoft.Extensions.Caching.Memory;
+
+namespace PixelServer.Controllers
+{
+    [Route("api/impressions")]
+    [ApiController]
+    public class TrackingController : ControllerBase
+    {
+        private readonly IMemoryCache _cache;
+        private readonly TelemetryClient telemetry;
```
Comment (mikeharder) [inline web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Rename to _telemetry for consistency

Comment (mikeharder) [inline web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Rename to `_telemetry` for consistency

File: web/pixel-server/PixelServer/Controllers/TrackingController.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+        public TrackingController(IMemoryCache memoryCache, TelemetryClient telemetry)
+        {
+            _cache = memoryCache;
+            this.telemetry = telemetry;
```
Comment (mikeharder) [inline web/pixel-server/PixelServer/Controllers/TrackingController.cs]: Remove `this.` for consistency

**Pull request merged**